designer/server/src/test/scala/pl/touk/nussknacker/test/config/WithRichDesignerConfig.scala

```diff
@@ -0,0 +1,116 @@
+package pl.touk.nussknacker.test.config
+
+import com.typesafe.config.{Config, ConfigFactory}
+import enumeratum.{Enum, EnumEntry}
+import io.restassured.specification.RequestSpecification
+import org.scalatest.Suite
+import pl.touk.nussknacker.engine.util.config.ScalaMajorVersionConfig
+import pl.touk.nussknacker.test.NuRestAssureExtensions
+import pl.touk.nussknacker.test.config.WithRichDesignerConfig.TestCategory
+import pl.touk.nussknacker.test.utils.DesignerTestConfigValidator
+
+trait WithRichDesignerConfig extends WithDesignerConfig {
+  this: Suite =>
+
+  validateConsistency()
+
+  override def designerConfig: Config = ScalaMajorVersionConfig.configWithScalaMajorVersion(
+    ConfigFactory.parseResources("config/rich/rich-streaming-use-case-designer.conf")
+  )
+
+  private def validateConsistency(): Unit = {
+    val configValidator = new DesignerTestConfigValidator(designerConfig)
+    val processingTypeWithCategories =
+      TestCategory.categoryByProcessingType.map { case (k, v) => (k.stringify, v.stringify) }
+    configValidator.validateTestDataWithDesignerConfFile(processingTypeWithCategories)
+  }
+
+}
+
+object WithRichDesignerConfig {
+  sealed trait TestProcessingType extends EnumEntry
+
+  object TestProcessingType extends Enum[TestProcessingType] {
+    case object Streaming1 extends TestProcessingType
+    case object Streaming2 extends TestProcessingType
+
+    override val values = findValues
+
+    implicit class ProcessingTypeStringify(processingType: TestProcessingType) {
+
+      def stringify: String = processingType match {
+        case TestProcessingType.Streaming1 => "streaming1"
+        case TestProcessingType.Streaming2 => "streaming2"
+      }
+
+    }
+
+    def categoryBy(processingType: TestProcessingType): TestCategory = {
+      processingType match {
+        case TestProcessingType.Streaming1 => TestCategory.Category1
+        case TestProcessingType.Streaming2 => TestCategory.Category2
+      }
+    }
+
+  }
+
+  sealed trait TestCategory extends EnumEntry
+
+  object TestCategory extends Enum[TestCategory] {
+    case object Category1 extends TestCategory
+    case object Category2 extends TestCategory
+
+    override val values = findValues
+
+    implicit class CategoryStringify(category: TestCategory) {
+
+      def stringify: String = category match {
+        case Category1 => "Category1"
```

Comment:
at the moment when you import `TestCategory._` you have the `CategoryStringify` ext available. We could use cats' Show, but we will need to import `cats.implicits._` (at least). We don't use cats much in tests, so I thought it'd be easier to use it without cats. 

So, I would not change it. But if you wish I can add the mechanism side by the one I proposed. WDYT?

Reply:
Ah, I thought that `Show` is just an interface with `show` method. I see that it is a type class. Let's ignore show. The root question was, why this extension class and pattern matching? Why not to just use standard polymorphism. I mean:
```
case object Category1 extends TestCategory {
  def stringify = "Category1"
}
```
:) I like extension classes but mainly when I can't just extend the interface

Reply:
in case of ADT it's just cleaner (at least looking from the FP perspective)

Reply:
Scala is mix of FP and OOP. Why in this situation do you think that haskel is better than java? :) Do you see any situation when polymorphism is worth usage or everwhere we should replace normal polymorphism with ad-hoc polymorphism with type classes?

I see a huge drawbacks of preferring pattern matching more than polymorphism - it doesn't scale. Let's say that former developer in the project replaced 10 methods by 10 pattern matching usages. As a current developer, you don't really know the contract of the interface because it is implicit, spread through all these places.
Let's say that you want to add another class implementing this interface. Because of the pattern matching you need to jump though all these places digging how this implicit interface looks like. For me pattern matching and extension classes are a kind of compromise but our first choice should be normal polymorphism.

Share me what's your opinion on that.

Reply:
I see the following pros:
* ADT is a simple structure with no additional methods that darken its shape
* extension is like a function - it can be locally defined (you don't extend the interface by adding additional methods to it), or can be added like a public method (near the ADT)
* because the functionality provided by an extension is in one place, it's easier to read and reason about
* when you add a new case to ADT, the compiler shows you the places where you should take into consideration the new code

cons:
* when you add a new case, you need to add some code in several places
* when you use the following convention for the ADT:
```scala
sealed trait X
object X {
  case object X1 extends X
  case object X2 extends X
}
```

and when you want to do sth like this:
```scala
X1.extenstionMethod()
```

you need to do:`import X._`. (probably you've already imported `import X.X1`

---

I don't want to write down all pros and cons of the OOP approach, but there is one thing in OOP that is abused - it's inheritance [;)](https://www.yegor256.com/2016/09/13/inheritance-is-procedural.html). And that's why I rather avoid it (for sure in case of ADT).